File: backend-old/app/routers/vpn.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from pydantic import BaseModel

from app.database import get_db
from app.models import User, Server, VPNConfig, Subscription, PlanType, SubStatus
from app.routers.auth import get_current_user_dep
from app.services.wireguard import (
    generate_keypair, build_client_config, ip_from_index,
    register_peer_on_server, remove_peer_from_server
)

router = APIRouter(prefix="/vpn", tags=["vpn"])

PLAN_MAX_CONFIGS = {PlanType.free: 1, PlanType.pro: 5, PlanType.business: 20}


class GenerateConfigRequest(BaseModel):
    server_id: int
    device_name: str = "Meu Dispositivo"


class ConfigOut(BaseModel):
    id: int
    server_name: str
    server_country: str
    country_code: str
    device_name: str
    vpn_ip: str
    is_active: bool


@router.get("/configs", response_model=list[ConfigOut])
async def list_configs(
    user: User = Depends(get_current_user_dep),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(VPNConfig, Server)
        .join(Server, VPNConfig.server_id == Server.id)
        .where(VPNConfig.user_id == user.id, VPNConfig.is_active == True)
    )
    rows = result.all()
    return [
        ConfigOut(
            id=cfg.id,
            server_name=srv.name,
            server_country=srv.country,
            country_code=srv.country_code,
            device_name=cfg.device_name,
            vpn_ip=cfg.vpn_ip,
            is_active=cfg.is_active,
        )
        for cfg, srv in rows
    ]


@router.post("/generate")
async def generate_config(
    body: GenerateConfigRequest,
    user: User = Depends(get_current_user_dep),
    db: AsyncSession = Depends(get_db),
):
    # Verificar plano e limites
    sub_result = await db.execute(
        select(Subscription).where(Subscription.user_id == user.id)
        .order_by(Subscription.id.desc())
    )
    sub = sub_result.scalar_one_or_none()
    user_plan = sub.plan if sub else PlanType.free

    if sub and sub.status not in (SubStatus.active, SubStatus.trialing):
        raise HTTPException(status_code=403, detail="Subscrição inativa. Atualize o pagamento.")

    config_count_result = await db.execute(
        select(func.count()).where(VPNConfig.user_id == user.id, VPNConfig.is_active == True)
    )
    config_count = config_count_result.scalar()
    if config_count >= PLAN_MAX_CONFIGS[user_plan]:
        raise HTTPException(
            status_code=403,
            detail=f"Limite de {PLAN_MAX_CONFIGS[user_plan]} configurações para o plano {user_plan.value}. Faça upgrade."
        )

    # Buscar servidor
    srv_result = await db.execute(select(Server).where(Server.id == body.server_id, Server.is_active == True))
    server = srv_result.scalar_one_or_none()
    if not server:
        raise HTTPException(status_code=404, detail="Servidor não encontrado")

    # Verificar acesso ao servidor pelo plano
    plan_order = {PlanType.free: 0, PlanType.pro: 1, PlanType.business: 2}
    if plan_order[user_plan] < plan_order[server.min_plan]:
        raise HTTPException(status_code=403, detail="Faça upgrade para aceder a este servidor.")

    # Gerar índice único para o IP do peer (total de peers já criados + 1)
    total_result = await db.execute(select(func.count()).select_from(VPNConfig))
    peer_index = (total_result.scalar() or 0) + 1
    vpn_ip = ip_from_index(peer_index)

    # Gerar chaves — a privada NÃO é armazenada
    private_key, public_key = generate_keypair()

    # Registar peer no servidor VPN
    success = await register_peer_on_server(server.agent_url, public_key, vpn_ip)
    if not success:
        raise HTTPException(status_code=503, detail="Erro ao registar no servidor VPN. Tente novamente.")

    # Persistir apenas a chave pública
    config = VPNConfig(
        user_id=user.id,
        server_id=server.id,
        wg_public_key=public_key,
        vpn_ip=vpn_ip,
        device_name=body.device_name,
    )
    db.add(config)
    server.active_peers += 1
    await db.commit()

    # Montar e retornar o ficheiro .conf (privada está aqui e é descartada após envio)
    conf_content = build_client_config(
        private_key=private_key,
        client_vpn_ip=vpn_ip,
        server_public_key=server.wg_public_key,
        server_endpoint=server.ip,
        server_port=server.wg_port,
    )

    from fastapi.responses import Response
    filename = f"pontevpn-{server.country_code.lower()}-{body.device_name.replace(' ', '_').lower()}.conf"
    return Response(
        content=conf_content,
        media_type="text/plain",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.delete("/configs/{config_id}")
async def revoke_config(
    config_id: int,
    user: User = Depends(get_current_user_dep),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(VPNConfig, Server)
        .join(Server)
        .where(VPNConfig.id == config_id, VPNConfig.user_id == user.id)
    )
    row = result.one_or_none()
    if not row:
        raise HTTPException(status_code=404, detail="Configuração não encontrada")

    cfg, srv = row
    await remove_peer_from_server(srv.agent_url, cfg.wg_public_key)
    cfg.is_active = False
    srv.active_peers = max(0, srv.active_peers - 1)
    await db.commit()
    return {"message": "Configuração revogada"}

File: backend/src/handlers/vpn.rs
use axum::{
    extract::State,
    Json,
    http::{StatusCode, header},
    response::{IntoResponse, Response},
};
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use crate::models::{User, Server, VPNConfig, Subscription, PlanType, SubStatus};
use chrono::Utc;
use crate::middleware::auth::ClaimsExtractor;
use crate::services::wireguard::WireGuardService;

#[derive(Deserialize)]
pub struct GenerateConfigRequest {
    pub server_id: i32,
    pub device_name: Option<String>,
}

#[derive(Serialize)]
pub struct ConfigOut {
    pub id: i32,
    pub server_name: String,
    pub server_country: String,
    pub country_code: String,
    pub device_name: String,
    pub vpn_ip: String,
    pub is_active: bool,
}

pub async fn list_configs(
    State(pool): State<PgPool>,
    ClaimsExtractor(claims): ClaimsExtractor,
) -> Json<Vec<ConfigOut>> {
    let user_id: i32 = claims.sub.parse().unwrap();

    let rows = sqlx::query!(
        r#"
        SELECT c.id, c.device_name, c.vpn_ip, c.is_active, s.name as server_name, s.country, s.country_code
        FROM vpn_configs c
        JOIN servers s ON c.server_id = s.id
        WHERE c.user_id = $1 AND c.is_active = true
        "#,
        user_id
    )
    .fetch_all(&pool)
    .await
    .unwrap();

    let result = rows
        .into_iter()
        .map(|r| ConfigOut {
            id: r.id,
            server_name: r.server_name,
            server_country: r.country,
            country_code: r.country_code,
            device_name: r.device_name,
            vpn_ip: r.vpn_ip,
            is_active: r.is_active,
        })
        .collect();

    Json(result)
}

pub async fn generate_config(
    State(pool): State<PgPool>,
    ClaimsExtractor(claims): ClaimsExtractor,
    Json(payload): Json<GenerateConfigRequest>,
) -> impl IntoResponse {
    let user_id: i32 = claims.sub.parse().unwrap();

    // 1. Verificar plano e limites
    let sub = sqlx::query_as!(
        Subscription,
        "SELECT * FROM subscriptions WHERE user_id = $1 ORDER BY id DESC LIMIT 1",
        user_id
    )
    .fetch_optional(&pool)
    .await
    .unwrap();

    let user = sqlx::query_as!(User, "SELECT * FROM users WHERE id = $1", user_id)
        .fetch_one(&pool)
        .await
        .unwrap();

    let user_plan = sub.as_ref().map(|s| &s.plan).unwrap_or(&PlanType::Free);
    let sub_status = sub.as_ref().map(|s| &s.status).unwrap_or(&SubStatus::Active);

    // Permite acesso se: Whitelisted OU Trial ativo OU Subscrição ativa
    let has_access = user.is_whitelisted || 
                     (user.trial_ends_at.is_some() && user.trial_ends_at.unwrap() > Utc::now()) || 
                     (*sub_status == SubStatus::Active || *sub_status == SubStatus::Trialing);

    if !has_access {
        return (StatusCode::FORBIDDEN, "Acesso expirado. Adquira um plano para continuar.").into_response();
    }

    let config_count = sqlx::query!("SELECT count(*) FROM vpn_configs WHERE user_id = $1 AND is_active = true", user_id)
        .fetch_one(&pool)
        .await
        .unwrap()
        .count
        .unwrap();

    let max_configs = match user_plan {
        PlanType::Free => 1,
        PlanType::Pro => 5,
        PlanType::Business => 20,
    };

    if config_count >= max_configs {
        return (StatusCode::FORBIDDEN, format!("Limite de {} atingido", max_configs)).into_response();
    }

    // 2. Buscar servidor
    let server = sqlx::query_as!(Server, "SELECT * FROM servers WHERE id = $1 AND is_active = true", payload.server_id)
        .fetch_optional(&pool)
        .await
        .unwrap();

    let server = match server {
        Some(s) => s,
        None => return (StatusCode::NOT_FOUND, "Servidor não encontrado").into_response(),
    };

    // 3. Gerar chaves e IP
    let (priv_key, pub_key) = WireGuardService::generate_keypair();
    let total_configs = sqlx::query!("SELECT count(*) FROM vpn_configs")
        .fetch_one(&pool)
        .await
        .unwrap()
        .count
        .unwrap();
    let vpn_ip = WireGuardService::ip_from_index(total_configs as i32 + 1);

    // 4. Registrar no servidor
    let secret = std::env::var("VPN_SERVERS_API_SECRET").unwrap_or_default();
    let success = WireGuardService::register_peer_on_server(&server.agent_url, &pub_key, &vpn_ip, &secret).await;

    if let Ok(true) = success {
        let device_name = payload.device_name.unwrap_or_else(|| "Meu Dispositivo".into());
        
        sqlx::query!(
            "INSERT INTO vpn_configs (user_id, server_id, wg_public_key, vpn_ip, device_name) VALUES ($1, $2, $3, $4, $5)",
            user_id, server.id, pub_key, vpn_ip, device_name
        )
        .execute(&pool)
        .await
        .unwrap();

        let conf = WireGuardService::build_client_config(&priv_key, &vpn_ip, &server.wg_public_key, &server.ip, server.wg_port);
        
        let filename = format!("pontevpn-{}.conf", server.country_code.to_lowercase());
        
        Response::builder()
            .header(header::CONTENT_TYPE, "text/plain")
            .header(header::CONTENT_DISPOSITION, format!("attachment; filename=\"{}\"", filename))
            .body(conf.into())
            .unwrap()
            .into_response()
    } else {
        (StatusCode::SERVICE_UNAVAILABLE, "Erro ao registrar peer").into_response()
    }
}
